nvidia-container-toolkit: do not shadow docker executable

ereslibre opened the pull request:
## Description of changes

At this time, the nvidia-container-toolkit derivation installs a docker executable that shadows the main one, and that is not thought to forward commands to the original docker command, causing issues to users when the `nvidia-container-toolkit` is in scope and they try to call to `docker`.

Fixes: https://github.com/NixOS/nixpkgs/issues/293857

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ereslibre committed:
nvidia-container-toolkit: do not shadow docker executable

At this time, the nvidia-container-toolkit derivation installs a
docker executable that shadows the main one, and that is not thought
to forward commands to the original docker command, causing issues to
users when the `nvidia-container-toolkit` is in scope and they try to
call to `docker`.

SomeoneSerge commented on pkgs/by-name/nv/nvidia-container-toolkit/package.nix:
```diff
@@ -56,6 +56,12 @@ buildGoModule rec {
     ./0001-Add-dlopen-discoverer.patch
   ];
 
+  subPackages = [
+    "cmd/nvidia-ctk"
+    "cmd/nvidia-container-runtime"
+    "cmd/nvidia-container-runtime-hook"
+  ];
+
```
I'm stupid, what does this do?

ereslibre replied:
`nvidia-container-toolkit` has a lot of binaries in its tree. Some of them are under https://github.com/NVIDIA/nvidia-container-toolkit/tree/main/cmd. Others are under https://github.com/NVIDIA/nvidia-container-toolkit/tree/main/tools/container.

The last one is the one causing the shadowing, but in reality the main issue is that some of those despite being named `docker` or `crio` are not meant to shadow the original binaries, but just named like that to set up the corresponding runtime configuration.

I believe we don't need to expose that kind of tooling to the user, and that NixOS will properly configure the given runtime with regular NixOS options. This is an implementation detail in my opinion.

So with `subPackages` what we are telling `buildGoModule` is: instead of building all the binaries that you can find in this tree structure, just build `cmd/nvidia-ctk`, `cmd/nvidia-container-runtime` and `cmd/nvidia-container-runtime-hook`.

To be honest, I would even just try with `cmd/nvidia-ctk` and the start adding more binaries as we see are missing, but this was a way to start.

ereslibre replied:
This is the list of binaries produced by `nix-build -A nvidia-docker` before this change:

```
❯ tree result/bin
result/bin
├── containerd -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/containerd
├── crio -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/crio
├── docker -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/docker
├── nvidia-container-cli -> /nix/store/570j23jms28cjb603a7hy695nmibfi0n-libnvidia-container-1.9.0/bin/nvidia-container-cli
├── nvidia-container-runtime -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-container-runtime
├── nvidia-container-runtime.cdi -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-container-runtime.cdi
├── nvidia-container-runtime-hook -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-container-runtime-hook
├── nvidia-container-runtime.legacy -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-container-runtime.legacy
├── nvidia-container-toolkit -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-container-toolkit
├── nvidia-ctk -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-ctk
├── nvidia-docker -> /nix/store/lc86iqnadxhc54n34n4g9rd94li39ls8-nvidia-docker-2.5.0/bin/nvidia-docker
├── nvidia-toolkit -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-toolkit
└── toolkit -> /nix/store/9933i2f5wnj6f994isn242vdnbzdry62-container-toolkit-container-toolkit-1.15.0-rc.3/bin/toolkit
```

And after this change:

```
❯ tree result/bin
result/bin
├── nvidia-container-cli -> /nix/store/570j23jms28cjb603a7hy695nmibfi0n-libnvidia-container-1.9.0/bin/nvidia-container-cli
├── nvidia-container-runtime -> /nix/store/iw49hbrnnxg2pf1ynha35gn96mgpg4xx-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-container-runtime
├── nvidia-container-runtime-hook -> /nix/store/iw49hbrnnxg2pf1ynha35gn96mgpg4xx-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-container-runtime-hook
├── nvidia-container-toolkit -> /nix/store/iw49hbrnnxg2pf1ynha35gn96mgpg4xx-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-container-toolkit
├── nvidia-ctk -> /nix/store/iw49hbrnnxg2pf1ynha35gn96mgpg4xx-container-toolkit-container-toolkit-1.15.0-rc.3/bin/nvidia-ctk
└── nvidia-docker -> /nix/store/lc86iqnadxhc54n34n4g9rd94li39ls8-nvidia-docker-2.5.0/bin/nvidia-docker
```

SomeoneSerge replied:
So instead of moving runtime wrappers to a separate output or prefix, this removes the wrappers entirely. I've no idea if the wrappers still have any users, I think we should go with the moving instead

ereslibre replied:
Please, note that the wrapper is `nvidia-docker`. The rest are just helper binaries that configure the runtime after they are called: tools that perform imperative changes on the host to set up the runtime wrapper, but not the wrappers themselves.

ereslibre replied:
Marking this PR as draft until I have double checked everything I want.

ereslibre committed:
virtualisation/docker: fix nvidia wrapper

Although CDI should be used in order to not require container runtime
wrappers anymore, fix the nvidia-container-runtime integration with
Docker for cases when Docker < 25.

ereslibre committed:
WIP

ereslibre commented:
Ouch, deleted the branch by mistake. Opening another PR since I cannot reopen this one.
